api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java

```diff
@@ -748,28 +813,47 @@ public Long getCountOfPatients(String query, boolean includeVoided) {
 
 		return patients;
 	}
-
-    private LuceneQuery<PatientIdentifier> getPatientIdentifierLuceneQuery(String query, boolean includeVoided) {
-		query = removeIdentifierPadding(query);
+	private LuceneQuery<PatientIdentifier> getPatientIdentifierLuceneQuery(String query, List<PatientIdentifierType> identifierTypes, boolean matchExactly) {
+		LuceneQuery<PatientIdentifier> patientIdentifierLuceneQuery = getPatientIdentifierLuceneQuery(query, matchExactly);
+		for(PatientIdentifierType identifierType : identifierTypes) {
+			patientIdentifierLuceneQuery.include("identifierType.patientIdentifierTypeId", identifierType.getId());
+		}
+		patientIdentifierLuceneQuery.include("patient.isPatient", true);
+		patientIdentifierLuceneQuery.skipSame("patient.personId");
+		
+		return patientIdentifierLuceneQuery;
+	}
+	
+	private LuceneQuery<PatientIdentifier> getPatientIdentifierLuceneQuery(String paramQuery, boolean matchExactly) {
+		String query = removeIdentifierPadding(paramQuery);
 		List<String> tokens = tokenizeIdentifierQuery(query);
-	    query = StringUtils.join(tokens, " OR ");
-	    List<String> fields = new ArrayList<>();
-	    fields.add("identifierPhrase");
-	
-	    String matchMode = Context.getAdministrationService().getGlobalProperty(OpenmrsConstants.GLOBAL_PROPERTY_PATIENT_IDENTIFIER_SEARCH_MATCH_MODE);
-	    if (OpenmrsConstants.GLOBAL_PROPERTY_PATIENT_SEARCH_MATCH_START.equals(matchMode)) {
-		    fields.add("identifierStart");
-	    } else if (OpenmrsConstants.GLOBAL_PROPERTY_PATIENT_SEARCH_MATCH_ANYWHERE.equals(matchMode)) {
-		    fields.add("identifierAnywhere");
-	    }
-	    LuceneQuery<PatientIdentifier> luceneQuery = LuceneQuery.newQuery(PatientIdentifier.class, sessionFactory.getCurrentSession(), query, fields);
+		query = StringUtils.join(tokens, " OR ");
+		List<String> fields = new ArrayList<>();
+		fields.add("identifierPhrase");
+		fields.add("identifierType");
+		String matchMode = Context.getAdministrationService()
+			.getGlobalProperty(OpenmrsConstants.GLOBAL_PROPERTY_PATIENT_IDENTIFIER_SEARCH_MATCH_MODE);
+		if (matchExactly) {
+			fields.add("identifierExact");
+		}
+		else if (OpenmrsConstants.GLOBAL_PROPERTY_PATIENT_SEARCH_MATCH_START.equals(matchMode)) {
+			fields.add("identifierStart");
+		} 
+		else if (!matchExactly || OpenmrsConstants.GLOBAL_PROPERTY_PATIENT_SEARCH_MATCH_ANYWHERE.equals(matchMode)) {
```

Comment:
matchExactly can never be true at this point.

Reply:
But it can be false. And if it is false this if statement will perhaps never be executed and no field is added. If I drop the !matchExactly the following test case is failing:


https://github.com/fruether/openmrs-core/blob/06287dee0a0b2d8187cd479c2833c8308690a14a/api/src/test/java/org/openmrs/api/PatientServiceTest.java#L474

Can you confirm that it is intended that this case fails? 

I never introduced these global properties that are requested but I have to live with them now...